Integrate widget framework

rhysyngsun opened the pull request:
#### Pre-Flight checklist

- [x] Migrations
  - [x] Migration is backwards-compatible with current production code
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
Fixes #1441 

#### What's this PR do?
Adds the widget framework

#### How should this be manually tested?

- Rebuild the app `docker-compose build`
- Migrate the app `docker-compose up`
- Verify you can view channels and posts without errors
- Run `./manage.py backpopulate_channel_widget_lists` to generate default widgets based off the channel descriptions
- Test modifying the widget list as a moderator user via the api:

API:

- GET `/api/v0/widget-lists/:id/`
- PATCH `/api/v0/widget-lists/:id/`

Example curl modifying the widget list (you'll need to add session info or comment out `permission_classes` on the view):

```
curl --header "Content-Type: application/json" \
     --request PATCH \
     --data '{"widgets":[{"widget_type": "Markdown", "title": "Test Title", "configuration": {"source": "*markdown*"}}]}' \
     http://od.dev.local:8063/api/v0/widget-lists/1/
```

noisecapella commented:
The CSS looks pretty bad: 
![screenshot from 2018-12-03 11-10-33](https://user-images.githubusercontent.com/863262/49389168-a5908d80-f6f3-11e8-8231-43f3742cbcce.png)

Feel free to ignore if it's handled in a future PR

noisecapella commented:
I'm seeing a 403 error when I try to edit the channel due to a lack of that permission. Is that in a future PR or is there some configuration I should do on my end?

noisecapella commented on channels/api.py:
```diff
@@ -385,6 +386,17 @@ def get_post_type(*, text, url, article_content):
     return LINK_TYPE_SELF
 
 
+def get_admin_api():
```
This is useful, not sure why we didn't do this sooner

noisecapella commented on channels/management/commands/backpopulate_channel_widget_lists.py:
```diff
@@ -0,0 +1,52 @@
+    help = "Create Channel WidgetList records"
+
+    def handle(self, *args, **options):
+        """Adds the indexing user as the moderator on all channels"""
```
Is this copied/pasted from another management command?

noisecapella commented on open_discussions/urls.py:
```diff
@@ -32,6 +32,7 @@
     url(r"", include("notifications.urls")),
     url(r"", include("embedly.urls")),
     url(r"", include("search.urls")),
+    url(r"", include("open_widget_framework.urls")),
```
There's a feature flag for the UI but not for the backend. I haven't looked at the backend code, do you think it's ready for release, or at least will deny non-admin users access appropriately?

rhysyngsun replied:
Looks like the urls to modify widgets is protected but not the one to modify lists is not. I think I'd like to actually fix this in an upstream PR.

rhysyngsun commented on channels/api.py:
```diff
@@ -471,7 +483,13 @@ def create_channel(
 
-        Channel.objects.create(name=name, membership_is_managed=membership_is_managed)
+        # create an empty widget list for new channels
+        widget_list = WidgetList.objects.create()
```
I agree, in fact, I've already addressed this in my [other PR for article work](https://github.com/mitodl/open-discussions/pull/1517/files#diff-6b4e079385252444280613805ab05474R515), because even with a transaction here, we can still end up in an inconsistent state because reddit doesn't get rolled back. Problem then, is you'll get an integrity error on the channel write that preempts the current reddit duplicate channel error we're relying on, so **that** then needs to get remapped.

Since I've already done all that work in that other PR I'd suggest either

a) deferring adding this transaction and just rebasing the other PR once this merges
b) block merging this PR until the other one is reviewed and merged 

Thoughts?

rhysyngsun commented:
@noisecapella fixed the permission issue and addressed feedback to the degree I can, I think another upstream change is required to properly secure this though.

rhysyngsun commented:
@noisecapella I think I've addressed the feedback and also fixed that lack of perm creation on the api I mentioned to you

rhysyngsun commented:
@noisecapella all set with the form specs now under `"available_widgets"`

noisecapella commented:
Thanks, can you rebase?

rhysyngsun commented:
@noisecapella rebased

noisecapella commented:
Looks like there's a stray `<<<<<` in there which the linting is complaining about

rhysyngsun commented:
@noisecapella fixed

noisecapella commented on widgets/serializers/react_fields.py:
```diff
@@ -0,0 +1,83 @@
+        """Returns the field properties"""
+        return {
+            **super()._get_props(),
+            "maxLength": self.max_length or "",
```
Can we change to `max_length` to match the rest of our APIs?

noisecapella commented on widgets/serializers/react_fields.py:
```diff
@@ -0,0 +1,83 @@
+        return {
+            **super()._get_props(),
+            "maxLength": self.max_length or "",
+            "minLength": self.min_length or "",
```
Can we change to `min_length`?

noisecapella commented on widgets/serializers/react_fields.py:
```diff
@@ -0,0 +1,83 @@
+        """Returns the field properties"""
+        return {
+            **super()._get_props(),
+            "maxLength": self.max_length or "",
```
Can we change these two to `max_length` and `min_length`?

noisecapella commented on widgets/urls.py:
```diff
@@ -0,0 +1,13 @@
+from widgets.views import WidgetListViewSet
+
+router = routers.DefaultRouter()
+router.register(r"widget-lists", WidgetListViewSet, basename="widget_list")
```
Can we change to `widget_lists` here, since we use underscores elsewhere in our REST API?

rhysyngsun commented on channels/api.py:
```diff
@@ -471,7 +485,16 @@ def create_channel(
+            widget_list=widget_list,
+        )
+
+        moderator_group = get_role_model(name, ROLE_MODERATORS).group
```
Hmmm, that's doable, but a little tricky, I think for now we probably won't worry about it if that's ok, but I'll open a ticket to look into it.

rhysyngsun replied:
#1573

noisecapella replied:
:+1:

rhysyngsun commented on channels/management/commands/backpopulate_channel_widget_lists.py:
```diff
@@ -0,0 +1,57 @@
+    def handle(self, *args, **options):
+        """Adds WidgetLists to existing channels"""
+        api = get_admin_api()
+        for channel_id in Channel.objects.values_list("id", flat=True).iterator():
```
This may have originally been the entire record, so it's probably unnecessary now.

noisecapella commented:
Looks like it's complaining about a black formatter warning
